builtin/logical/ssh/path_issue.go

```diff
@@ -0,0 +1,207 @@
+package ssh
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/hashicorp/go-secure-stdlib/strutil"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+	"golang.org/x/crypto/ssh"
+)
+
+func pathIssue(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "issue/" + framework.GenericNameWithAtRegex("role"),
+
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback: b.pathIssue,
+			},
+		},
+		Fields: map[string]*framework.FieldSchema{
+			"role": {
+				Type:        framework.TypeString,
+				Description: `The desired role with configuration for this request.`,
+			},
+			"key_type": {
+				Type:        framework.TypeString,
+				Description: "TBD",
+				Required:    true,
+			},
+			"key_bits": {
+				Type:        framework.TypeInt,
+				Description: "TBD",
+			},
+			"ttl": {
+				Type: framework.TypeDurationSecond,
+				Description: `The requested Time To Live for the SSH certificate;
+sets the expiration date. If not specified
+the role default, backend default, or system
+default TTL is used, in that order. Cannot
+be later than the role max TTL.`,
+			},
+			"valid_principals": {
+				Type:        framework.TypeString,
+				Description: `Valid principals, either usernames or hostnames, that the certificate should be signed for.`,
+			},
+			"cert_type": {
+				Type:        framework.TypeString,
+				Description: `Type of certificate to be created; either "user" or "host".`,
+				Default:     "user",
+			},
+			"key_id": {
+				Type:        framework.TypeString,
+				Description: `Key id that the created certificate should have. If not specified, the display name of the token will be used.`,
+			},
+			"critical_options": {
+				Type:        framework.TypeMap,
+				Description: `Critical options that the certificate should be signed for.`,
+			},
+			"extensions": {
+				Type:        framework.TypeMap,
+				Description: `Extensions that the certificate should be signed for.`,
+			},
+		},
+		HelpSynopsis:    "TBD - HelpSynopsis",
+		HelpDescription: "TBD - HelpDescription",
+	}
+}
+
+func (b *backend) pathIssue(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role").(string)
+	role, err := b.getRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("unknown role: %s", roleName)), nil
+	}
+
+	if role.KeyType != "ca" {
+		return logical.ErrorResponse("role key type \"%s\" not allowed to issue key pairs", role.KeyType), nil
+	}
+
+	keyType := data.Get("key_type").(string)
+	keyBits := data.Get("key_bits").(int)
+
+	keyTypeLengths, keyPresent := role.AllowedUserKeyTypesLengths[keyType]
```

Comment:
I think you'll end up wanting to hoist this to a package level internal constant:

https://github.com/hashicorp/vault/blob/7c72d9d3d6f127ff368d6f47634684c6fb60cb22/builtin/logical/ssh/path_sign.go#L476-L481 

There's a series of aliases we allow (for better usability) and you've gotta check them all (for compatibility with the existing validatiog logic). 

And just to clarify this, they are parameters/field in the allowed map, not in the requested params. There you can keep it simple and align with the PKI endpoint (rsa/ec/ed25519).

Reply:
Additionally, this check (lines 89 through 102) needs to be conditional on whether or not the map is even present:

https://github.com/hashicorp/vault/blob/7c72d9d3d6f127ff368d6f47634684c6fb60cb22/builtin/logical/ssh/path_sign.go#L450

If the map isn't present, you should allow all key types/bits. :-)

Well, I'd deny RSA with bits < 2048, but yeah.... Oh and Go doesn't really have DSA support any more, so please don't implement it either :D It is supported on signing since users can generate DSA keys, but its generally considered weak now.

Reply:
https://github.com/hashicorp/vault/blob/7c72d9d3d6f127ff368d6f47634684c6fb60cb22/builtin/logical/ssh/path_sign.go#L476-L481


Is it okay to make this global variable as it can be mutated here?
https://github.com/hashicorp/vault/blob/d897316a9e337a2a065ac738c1240490f33eb2ed/builtin/logical/ssh/path_sign.go#L412-L414

Reply:
Ah sharp eyes, probably best not to mutate it then. I was just thinking that I was regretting writing such complex algorithmic verification...